File: 002_add_two_numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:

        carry = None

        el = l1.val + l2.val
        if (el >= 10):
            el -= 10
            carry = 1

        ans = ListNode(el)
        curr = ans

        l1 = l1.next
        l2 = l2.next

        while (l1 != None or l2 != None):

            if (l1 is None):

                el = None

                if (carry != None):
                    el = l2.val + carry
                    if (el >= 10):
                        carry = 1
                        el -= 10
                    else:
                        carry = None
                else:
                    el = l2.val

                curr.next = ListNode(el)

                l2 = l2.next
                curr = curr.next

            elif (l2 is None):
                el = None

                if (carry != None):
                    el = l1.val + carry
                    if (el >= 10):
                        carry = 1
                        el -= 10
                    else:
                        carry = None
                else:
                    el = l1.val

                curr.next = ListNode(el)

                l1 = l1.next
                curr = curr.next
            else:
                el = l1.val + l2.val

                if (carry != None):
                    el += carry
                    carry = None

                if (el >= 10):
                    carry = 1
                    el = el - 10
                curr.next = ListNode(el)

                curr = curr.next
                l1 = l1.next
                l2 = l2.next

        if (carry):
            curr.next = ListNode(carry)
        return ans

File: 002_add_two_numbers.ts
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

function addTwoNumbers(
    l1: ListNode | null,
    l2: ListNode | null
): ListNode | null {
    const ans = new ListNode();
    let curr = ans;
    let carry = false;

    while (l1 !== null && l2 !== null) {
        let temp = l1.val + l2.val;
        if (carry) {
            temp += 1;
        }
        carry = temp >= 10;
        curr.val = temp % 10;
        if (l1.next !== null && l2.next !== null) {
            curr.next = new ListNode();
            curr = curr.next;
        }
        l1 = l1.next;
        l2 = l2.next;
    }

    while (l1 !== null) {
        let temp = l1.val;
        if (carry) temp += 1;
        carry = temp >= 10;
        curr.next = new ListNode(temp % 10);
        curr = curr.next;
        l1 = l1.next;
    }

    while (l2 !== null) {
        let temp = l2.val;
        if (carry) temp += 1;
        carry = temp >= 10;
        curr.next = new ListNode(temp % 10);
        curr = curr.next;
        l2 = l2.next;
    }

    if (carry) {
        curr.next = new ListNode(1);
    }

    return ans;
}
